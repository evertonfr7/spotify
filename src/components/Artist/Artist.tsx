import Avatar from '../ui/Avatar'
import Typography from '../ui/Typography'
import { ArtistProps } from './Artist.types'

export function Artist({ id, href, name, image }: ArtistProps) {
  return (
    <a href={href}>
      <div className="flex items-center gap-2 md:gap-4 text-white" key={id}>
        <Avatar src={image} alt={name} ariaLabel={name} />
        <Typography className="font-normal">{name}</Typography>
      </div>
    </a>
  )
}
