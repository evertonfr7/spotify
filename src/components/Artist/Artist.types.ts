export interface ArtistProps {
  id: string
  href: string
  name: string
  image: string
}
