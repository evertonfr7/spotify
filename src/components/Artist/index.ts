export { Artist as default } from './Artist'
