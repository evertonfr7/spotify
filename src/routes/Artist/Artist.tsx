import Cover from '@/components/Cover'
import BaseNavigation from '@/components/layout/BaseNavigation'
import Avatar from '@/components/ui/Avatar'
import Button from '@/components/ui/Button'
import { ArrowLeft } from '@/components/ui/icons'
import Skeleton from '@/components/ui/Skeleton'
import Typography from '@/components/ui/Typography'
import { ALBUMS_PER_PAGE } from '@/constants'
import { getArtist } from '@/resources/http/api/queries/artist'
import { getArtistAlbuns } from '@/resources/http/api/queries/artist-albuns'
import { Album } from '@/resources/http/api/queries/artist-albuns/types'
import { useInfiniteQuery, useQuery } from '@tanstack/react-query'
import { useCallback } from 'react'

const LoadingAlbums = () => {
  return Array.from({ length: ALBUMS_PER_PAGE }).map((_, index) => (
    <div className="flex gap-2 md:gap-4 items-center" key={index}>
      <div>
        <Skeleton className="w-[48px] h-[48px] md:w-[72px] md:h-[72px]" />
      </div>
      <div>
        <Skeleton className="w-[150px] h-[18px] mb-1" />
        <Skeleton className="w-[100px] h-[18px]" />
      </div>
    </div>
  ))
}

export function Artist(): JSX.Element {
  const id = window.location.pathname.split('/').pop()

  const limit = ALBUMS_PER_PAGE

  const { data: artist, isLoading: isLoadingArtist } = useQuery({
    queryKey: ['artist', id],
    queryFn: () => getArtist({ id: id as string }),
  })

  const { data, isLoading, hasNextPage, fetchNextPage, isFetchingNextPage } =
    useInfiniteQuery({
      queryKey: ['artist-albuns', `limit=${limit}`],
      queryFn: ({ pageParam = 0 }) =>
        getArtistAlbuns({ id: id as string, limit, offset: pageParam }),
      initialPageParam: 0,
      getNextPageParam: (lastPage) => lastPage.data.offset + limit,
    })

  const loadMore = useCallback(() => {
    if (hasNextPage && !isFetchingNextPage) {
      fetchNextPage()
    }
  }, [hasNextPage, isFetchingNextPage, fetchNextPage])

  return (
    <BaseNavigation>
      <div className="flex items-center justify-between mb-4 md:mb-8">
        <div
          className="flex items-center gap-2"
          role="button"
          tabIndex={0}
          aria-label="Voltar"
          onKeyDown={() => window.history.back()}
          onClick={() => window.history.back()}
        >
          <ArrowLeft />
          {isLoadingArtist ? (
            <Skeleton className="w-[90px] h-[18px] md:w-[100px] md:h-[20px]" />
          ) : (
            <Typography className="font-bold">{artist?.data?.name}</Typography>
          )}
        </div>

        {isLoadingArtist ? (
          <Skeleton
            circle
            className="w-[48px] h-[48px] md:w-[64px] md:h-[64px]"
          />
        ) : (
          <Avatar
            src={artist?.data?.images[0].url}
            alt={artist?.data?.name}
            ariaLabel={artist?.data?.name}
          />
        )}
      </div>
      <div className="flex flex-col gap-4 overflow-y-auto h-[calc(100dvh-122px)] pb-4">
        {isLoading ? (
          <LoadingAlbums />
        ) : (
          <>
            {data?.pages?.map((page, index) => (
              <div key={index} className="flex flex-col gap-4">
                {page?.data?.items?.map((album: Album) => (
                  <Cover
                    type="album"
                    key={album.id}
                    id={album.id}
                    title={album.name}
                    subtitle={`${album.release_date.split('-')[2]}/${album.release_date.split('-')[1]}/${album.release_date.split('-')[0]}`}
                    image={album.images[0].url}
                  />
                ))}
              </div>
            ))}
            {hasNextPage && (
              <Button
                className="w-[fit-content]"
                disabled={isFetchingNextPage}
                onClick={loadMore}
              >
                Carregar mais
              </Button>
            )}
          </>
        )}
      </div>
    </BaseNavigation>
  )
}
