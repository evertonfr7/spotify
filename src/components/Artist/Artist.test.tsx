import { render, screen } from '@testing-library/react'
import { Artist } from './Artist'
import '@testing-library/jest-dom'

describe('Artist Component', () => {
  const mockProps = {
    id: 'artist-1',
    href: '/artist/1',
    name: 'Test Artist',
    image: '/images/test-artist.jpg',
  }

  test('renders the artist name correctly', () => {
    render(<Artist {...mockProps} />)
    const artistName = screen.getByText(mockProps.name)
    expect(artistName).toBeInTheDocument()
  })

  test('renders the artist avatar with correct alt text and src', () => {
    render(<Artist {...mockProps} />)
    const avatarImage = screen.getByRole('img', { name: mockProps.name })
    expect(avatarImage).toHaveAttribute('src', mockProps.image)
    expect(avatarImage).toHaveAttribute('alt', mockProps.name)
  })

  test('renders the link with correct href', () => {
    render(<Artist {...mockProps} />)
    const linkElement = screen.getByRole('link')
    expect(linkElement).toHaveAttribute('href', mockProps.href)
  })

  test('applies the correct class names to the container', () => {
    render(<Artist {...mockProps} />)
    const container = screen.getByRole('link').firstChild
    expect(container).toHaveClass('flex items-center gap-2 md:gap-4 text-white')
  })
})
