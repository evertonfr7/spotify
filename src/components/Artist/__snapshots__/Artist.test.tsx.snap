// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Artist Component renders the artist with correct snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <a
        href="/artist/1"
      >
        <div
          class="flex items-center gap-2 md:gap-4 text-white"
        >
          <img
            alt="Test Artist"
            aria-label="Test Artist"
            class="rounded-full aspect-auto object-cover flex items-center justify-center font-default font-normal text-black text-[24px] md:text-[42px] w-[48px] h-[48px] md:w-[64px] md:h-[64px]"
            height="128"
            role="img"
            src="/images/test-artist.jpg"
            width="128"
          />
          <p
            class="font-default text-[12px]/[18px] text-white md:text-[14px]/[20px] font-normal"
          >
            Test Artist
          </p>
        </div>
      </a>
    </div>
  </body>,
  "container": <div>
    <a
      href="/artist/1"
    >
      <div
        class="flex items-center gap-2 md:gap-4 text-white"
      >
        <img
          alt="Test Artist"
          aria-label="Test Artist"
          class="rounded-full aspect-auto object-cover flex items-center justify-center font-default font-normal text-black text-[24px] md:text-[42px] w-[48px] h-[48px] md:w-[64px] md:h-[64px]"
          height="128"
          role="img"
          src="/images/test-artist.jpg"
          width="128"
        />
        <p
          class="font-default text-[12px]/[18px] text-white md:text-[14px]/[20px] font-normal"
        >
          Test Artist
        </p>
      </div>
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
